diff --git a/src/burner/libretro/Makefile b/src/burner/libretro/Makefile
old mode 100644
new mode 100755
index 0f097af..8a96517
--- a/src/burner/libretro/Makefile
+++ b/src/burner/libretro/Makefile
@@ -359,6 +359,22 @@ else ifeq ($(platform), classic_armv7_a7)
 	CFLAGS += $(PLATFORM_FLAGS)
 	CXXFLAGS += $(PLATFORM_FLAGS)
 
+# target platform tg5040 trimUI brick
+else ifneq (,$(findstring tg5040,$(platform)))
+	TARGET := $(TARGET_NAME)_libretro.so
+	CXX=$(CROSS_COMPILE)g++
+	fpic := -fPIC
+	SHARED := -shared -Wl,-no-undefined -Wl,--version-script=$(VERSION_SCRIPT)
+	PLATFORM_FLAGS := -O3 -Ofast -DNDEBUG=1 -fomit-frame-pointer -ffast-math -mtune=cortex-a53 -mcpu=cortex-a53 -march=armv8-a+simd
+	USE_EXPERIMENTAL_FLAGS = 0
+	HAVE_NEON = 1
+	USE_CYCLONE = 1
+	ARCH = arm
+	ENDIANNESS_DEFINES := -DLSB_FIRST
+	LDFLAGS += -lstdc++ -static-libgcc -static-libstdc++ -lpthread -lm -flto
+	CFLAGS += $(PLATFORM_FLAGS)
+	CXXFLAGS += $(PLATFORM_FLAGS)
+	
 # (armv8 a35, hard point, neon based) ###
 # Playstation Classic
 else ifeq ($(platform), classic_armv8_a35)
diff --git a/src/burner/libretro/libretro.cpp b/src/burner/libretro/libretro.cpp
index 265ce39..7f0ef60 100644
--- a/src/burner/libretro/libretro.cpp
+++ b/src/burner/libretro/libretro.cpp
@@ -79,7 +79,7 @@ INT32 nAudSegLen = 0;
 
 static UINT8* pVidImage              = NULL;
 static bool bVidImageNeedRealloc     = false;
-static bool bRotationDone            = false;
+static bool bRotationDone            = true;
 static int16_t *pAudBuffer           = NULL;
 static char text_missing_files[2048] = "";
 
@@ -1565,8 +1565,46 @@ void retro_run()
 		// current frame will be corrupted, let's dupe instead
 		pBurnDraw = NULL;
 	}
-
-	video_cb(pBurnDraw, nGameWidth, nGameHeight, nBurnPitch);
+    // get flags
+    UINT32 flags = BurnDrvGetFlags();
+
+    if (pBurnDraw && flags & BDF_ORIENTATION_VERTICAL) {
+        int y=0;
+
+        UINT8 *pTMP = (UINT8*)malloc(nGameWidth * nGameHeight * nBurnBpp);
+        memcpy(pTMP, pBurnDraw, nGameWidth * nGameHeight * nBurnBpp);
+
+        int newGameWidth = nGameHeight;
+        int newGameHeight = nGameWidth;
+        int newBurnPitch = newGameWidth * nBurnBpp;
+
+		if ((BurnDrvGetFlags() & BDF_ORIENTATION_VERTICAL))
+        while (y < newGameHeight) {
+            int x = 0;
+
+		    if ( flags & BDF_ORIENTATION_FLIPPED ) {
+                while (x < newGameWidth)
+                {
+                    for (int i = 0; i < nBurnBpp; i++)
+                        pBurnDraw[x * nBurnBpp + y * newBurnPitch + i] = pTMP[(nGameHeight - x - 1) * nBurnPitch + y * nBurnBpp + i];
+                    x++;
+                }
+            } else {
+                while (x < newGameWidth)
+                {
+                    for (int i = 0; i < nBurnBpp; i++)
+                        pBurnDraw[x * nBurnBpp + y * newBurnPitch + i] = pTMP[x * nBurnPitch + (nGameWidth - y - 1) * nBurnBpp + i];
+                    x++;
+                }
+            }
+            y++;
+        }
+        free(pTMP);
+	    video_cb(pBurnDraw, newGameWidth, newGameHeight, newBurnPitch);
+    } else {
+	    video_cb(pBurnDraw, nGameWidth, nGameHeight, nBurnPitch);
+    }
 
 	bool updated = false;
 	if (environ_cb(RETRO_ENVIRONMENT_GET_VARIABLE_UPDATE, &updated) && updated)
@@ -1584,7 +1622,7 @@ void retro_run()
 		// change orientation/geometry if vertical mode was toggled on/off
 		if (old_nVerticalMode != nVerticalMode)
 		{
-			SetRotation();
+			//SetRotation();
 			struct retro_system_av_info av_info;
 			retro_get_system_av_info(&av_info);
 			environ_cb(RETRO_ENVIRONMENT_SET_GEOMETRY, &av_info);
@@ -2211,7 +2249,7 @@ static bool retro_load_game_common()
 
 		// Initializing display, autorotate if needed
 		BurnDrvGetFullSize(&nGameWidth, &nGameHeight);
-		SetRotation();
+		//SetRotation();
 		BurnSetResolution(nNewWidth, nNewHeight);
 		SetColorDepth();
 
diff --git a/src/burner/libretro/retro_common.cpp b/src/burner/libretro/retro_common.cpp
index 0424dd2..caca6fb 100644
--- a/src/burner/libretro/retro_common.cpp
+++ b/src/burner/libretro/retro_common.cpp
@@ -134,16 +134,16 @@ static struct retro_core_option_v2_definition var_fbneo_resolution = {
 	{
 		{ "640x480",        NULL },
 		{ "800x600",        NULL },
-		{ "1024x768",       NULL },
+		{ "1024x768",       NULL },
 		{ "1080x810",       NULL },
 		{ "1280x960",       NULL },
 		{ "1440x1080",      NULL },
 		{ "1600x1200",      NULL },
 		{ "1920x1440",      NULL },
 		{ "2160x1620",      NULL },
-		{ "2880x2160",      NULL },
+		{ "2880x2160",      NULL },
 	},
-	"640x480"
+	"800x600"
 };
 static struct retro_core_option_v2_definition var_fbneo_fixed_frameskip = {
 	"fbneo-fixed-frameskip",
